### PR TITLE
DIT-292 Filter by Trader Name

### PR DESCRIPTION
**Pull request by ed0906**

### CODE REVIEW

File: app/uk/gov/hmrc/bindingtariffclassification/model/search/Search.scala [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  filter: Filter = Filter(),
+  sort: Sort = Sort()
+)
+
+case class Filter
+(
+  queueId: Option[String] = None,
+  assigneeId: Option[String] = None,
+  status: Option[String] = None,
+  reference: Option[String] = None,
```
Comment (ed0906) [inline app/uk/gov/hmrc/bindingtariffclassification/model/search/Search.scala]: Will remove this

File: app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala [hunk 2]
```diff
@@ -45,13 +46,42 @@ trait CaseRepository {
...
 
-  def get(searchBy: CaseParamsFilter, sortedBy: Option[CaseSort]): Future[Seq[Case]]
+  def get(search: Search): Future[Seq[Case]]
 
   def deleteAll(): Future[Unit]
 }
 
 @Singleton
-class CaseMongoRepository @Inject()(mongoDbProvider: MongoDbProvider, jsonMapper: JsonObjectMapper)
+class EncryptingMongoRepository @Inject()(repository: CaseMongoRepository, crypto: Crypto, appConfig: AppConfig) extends CaseRepository {
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: We have now two mongo concrete singleton implementations of the trait `CaseRepository`.
How do you tell the Play microservice that we want only one to run? 
I think we do not want both running

Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: I think the name `EncryptedMongoRepository` sounds better (in my opinion)

Comment (gokyo) [reply app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: Actually, it should be `EncryptedCaseMongoRepository`

File: test/unit/uk/gov/hmrc/bindingtariffclassification/repository/EncryptingMongoRepositoryTest.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import uk.gov.hmrc.bindingtariffclassification.crypto.Crypto
+import uk.gov.hmrc.bindingtariffclassification.model.Case
+import uk.gov.hmrc.bindingtariffclassification.model.search.Search
+import uk.gov.hmrc.play.test.UnitSpec
+
+import scala.concurrent.Future.successful
+
+class EncryptingMongoRepositoryTest extends UnitSpec with MockitoSugar with BeforeAndAfterEach {
+
+  private val encryptionEnabled = MongoEncryption(enabled = true, key = Some("1AW32543H!="))
```
[the remark this answers is not in the document]
Comment (ed0906) [reply test/unit/uk/gov/hmrc/bindingtariffclassification/repository/EncryptingMongoRepositoryTest.scala]: I just copied this from the old service test. I don't think the key field even gets used

File: app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala [hunk 4]
```diff
@@ -45,13 +46,42 @@ trait CaseRepository {
...
+
+  override def get(search: Search): Future[Seq[Case]] = {
+    repository.get(encrypt(search)).map(_.map(decrypt))
+  }
+
+  override def deleteAll(): Future[Unit] = repository.deleteAll()
+}
+
+@Singleton
+class CaseMongoRepository @Inject()(mongoDbProvider: MongoDbProvider, mapper: SearchMapper)
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: Do you think we should keep this class? Or is it adding only more work? We would need to maintain it, but we do not use it I believe

File: test/unit/uk/gov/hmrc/bindingtariffclassification/repository/EncryptingMongoRepositoryTest.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ed0906) [reply test/unit/uk/gov/hmrc/bindingtariffclassification/repository/EncryptingMongoRepositoryTest.scala]: Removed it

File: app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala (hunk 4 above)
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: we 100% need this class

File: app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala [hunk 5]
```diff
@@ -45,13 +46,42 @@ trait CaseRepository {
...
+  def get(search: Search): Future[Seq[Case]]
 
   def deleteAll(): Future[Unit]
 }
 
 @Singleton
-class CaseMongoRepository @Inject()(mongoDbProvider: MongoDbProvider, jsonMapper: JsonObjectMapper)
+class EncryptingMongoRepository @Inject()(repository: CaseMongoRepository, crypto: Crypto, appConfig: AppConfig) extends CaseRepository {
+  private def encrypt: Case => Case = { c: Case =>
+    if (appConfig.mongoEncryption.enabled) crypto.encrypt(c) else c
```
[the remark this answers is not in the document]
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: Good idea but we'd have to toggle it in a Guice Module which would be a bit awkward to test

[the remark this answers is not in the document]
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: I'll have a go

File: app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala [hunk 6]
```diff
@@ -23,18 +23,19 @@ import reactivemongo.api.indexes.Index
...
+import uk.gov.hmrc.bindingtariffclassification.model.search.Search
 import uk.gov.hmrc.bindingtariffclassification.model.{Case, MongoFormatters}
 import uk.gov.hmrc.mongo.ReactiveRepository
 import uk.gov.hmrc.mongo.json.ReactiveMongoFormats
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
-@ImplementedBy(classOf[CaseMongoRepository])
+@ImplementedBy(classOf[EncryptedMongoRepository])
```
Comment (gokyo) [inline app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: We should call it `EncryptedCaseMongoRepository`

File: app/uk/gov/hmrc/bindingtariffclassification/crypto/Crypto.scala [hunk 7]
```diff
@@ -27,10 +28,22 @@ class Crypto @Inject()(crypto: CompositeSymmetricCrypto) {
     applyCrypto(c) { s: String => crypto.encrypt(PlainText(s)).value }
   }
 
+  def encrypt(search: Search): Search = {
+    applyCrypto(search) { s: String => crypto.encrypt(PlainText(s)).value }
```
[the remark this answers is not in the document]
Comment (pllaurado) [reply app/uk/gov/hmrc/bindingtariffclassification/crypto/Crypto.scala]: Could we leave this one until further notice?

[the remark this answers is not in the document]
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/crypto/Crypto.scala]: Yep

File: app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (ed0906) [reply app/uk/gov/hmrc/bindingtariffclassification/repository/CaseRepository.scala]: Done

**Pull request merged**